### PR TITLE
Remove `LegacyDifficultyControlPoint` / `LegacyBpmMultiplier`

### PR DESCRIPTION
**Pull request by peppy**
As pointed out [in discussion](https://github.com/ppy/osu/issues/24708#issuecomment-1708062276), the flow of data surrounding these legacy pieces is ugly. It only got worse after difficulty data has been moved to hit objects themselves (with the worst code smell being `LegacyBpmMultiplier` existing in `HitObject`.

This PR removes all of this, instead locally computing the required precision-adjusted values where necessary. It also moves the still-required `GenerateTicks` to `DifficultyControlPoint` itself (we basically need to know if the BPM at point-in-time was `NaN`, but don't want to convey that in `TimingControlPoint`, so this is required to exist somewhere).

To make this work, I've also removed the precision limitations of `SliderVelocityMultiplier`. This doesn't seem to have an adverse effect on the editor luckily, so should play quite well.

This PR fixes a few issues which seemingly went silently unnoticed:

- `SliderVelocityMultiplier` precision was not transferred to `HitObject` when the [value is transferred](https://github.com/ppy/osu/blob/1a37543d288fa3b62d19a34d443ee97943920fbd/osu.Game/Beatmaps/Formats/LegacyBeatmapDecoder.cs#L113-L114)
- `LegacyBpmMultiplier` was not being transferred at all [in `OsuBeatmapConverter`](https://github.com/ppy/osu/blob/1a37543d288fa3b62d19a34d443ee97943920fbd/osu.Game.Rulesets.Osu/Beatmaps/OsuBeatmapConverter.cs#L37-L53), so it was not being used anywhere (potentially affected difficulty calculation and gameplay)

I haven't added test coverage for these two cases because I'm not sure where to start for that.

### CODE REVIEW

**Review by bdach — DISMISSED** [review]

**Comment by bdach** [comment]
Tried to do sheets for this one.

- osu!: https://docs.google.com/spreadsheets/d/17OJPDS7X1cNa_XNK5bAh_ybTtxOJtTeGUA8bRmLt-zE/edit#gid=1975777070
- taiko: https://docs.google.com/spreadsheets/d/1OYh7pg3vyOfcDhg5YKrLkYB8V4_mL5cX01Ugf26hMmg/edit#gid=495941023
- catch: https://docs.google.com/spreadsheets/d/1B0FElU20L02gebaIuByib3NK1jehbnLJ7Ety7OTtOGk/edit#gid=779945743
- mania: TBA (still generating)

The taiko changes are slightly worrying as I wouldn't expect anything to change there, but I have done zero investigation what that is so far. Could even be that the generation somehow didn't work correctly on my machine, so reader beware.

**Comment by smoogipoo** [comment]
Fwiw, taiko is where I'd expect any changes to happen, because the drumroll conversion code relies on super precise errors.

**Comment by peppy** [comment]
FWIW I'd expect there to be changes as I fixed several actual bugs. We'll need to confirm they are correct to our expectations as a result.

**Comment by bdach** [comment]
Well, there is at least one breakage on the mania side when compared with stable.

[/b/2364268](https://osu.ppy.sh/beatmapsets/1131887#osu/2364268) with DT, 10K:

https://github.com/ppy/osu/assets/20418176/e49fb0f0-6db6-4172-9fc2-949c1e2511d2

Appears to be caused by minuscule fluctuations of `beatLength` in the pattern generator:

```
$ diff /d/Open\ Source/osu-notes/24738/2364269_before.txt /d/Open\ Source/osu-notes/24738/2364269_after.txt
10,16c10,16
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 26279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 27779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 29279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 30029: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 31341: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 37904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 38091: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 26279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 27779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 29279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 30029: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 31341: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 37904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 38091: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
74,79c74,79
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71279: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71654: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72029: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72404: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72779: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 73904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71279: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71654: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72029: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72404: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72779: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 73904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
99,104c99,104
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86841: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 87779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89841: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 90779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86841: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 87779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89841: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 90779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
114,120c114,120
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 26279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 27779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 29279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 30029: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 31341: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 37904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 38091: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 26279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 27779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 29279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 30029: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 31341: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 37904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 38091: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
178,183c178,183
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71279: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71654: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72029: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72404: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72779: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 73904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325684
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71279: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 71654: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72029: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72404: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 72779: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 73904: DistanceObjectPatternGenerator calculated beatLength = 249.99999046325686
203,208c203,208
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86841: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 87779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89279: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89841: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
< osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 90779: DistanceObjectPatternGenerator calculated beatLength = 499.9999809265137
\ No newline at end of file
---
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 86841: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 87779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89279: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 89841: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
> osu.Game.Rulesets.Objects.Legacy.Osu.ConvertSlider @ 90779: DistanceObjectPatternGenerator calculated beatLength = 499.99998092651373
\ No newline at end of file
```

The underlying cause is probably that taking the reciprocal of `LegacyBpmMultiplier` as `SliderVelocity` is not lossless.

Haven't looked at the taiko cases yet, but I'm not sure doing so is worth the non-trivial effort anymore given that we know that this change technically breaks stuff now... (getting to the bottom of the above took me ~2hrs already).

**Comment by peppy** [comment]
I've adjusted things to the point it looks to match your output (to within `ToString()` imprecisions as discussed).

I think we can probably use this as a way to test github actions diffcalc runs, so hold off on the manual run until I give that a try (@smoogipoo has been working on updating the action and I'll re-deploy my local runner).

**Comment by bdach** [comment]
Map mentioned above looks to be back to the reference state now.

As discussed elsewhere, going to re-run the sheets overnight to verify.

**Comment by bdach** [comment]
New sheets:

- osu!: https://docs.google.com/spreadsheets/d/1cEYpFThpsvVqNbIsltieCQ5d0n4HT7X5b08Y-u9dga4/edit#gid=120974302
- taiko: https://docs.google.com/spreadsheets/d/1WbdHXs-NCiYpg4Pf_KAmitkiUURFvxDmgHq_qULczUQ/edit#gid=671685342
- catch: https://docs.google.com/spreadsheets/d/1dIgmvKQh9SxC6JPlAUw02C_l4ShqFzBfs8Vr5l-fGTU/edit#gid=250029157
- mania: https://docs.google.com/spreadsheets/d/1TayrD1bQbe_u4F8g1_F5cBoDvpG4Ok9qN0Poxd4QxyA/edit#gid=1429609841

Things look clean, so no further objections from me.

**Review by bdach — DISMISSED** [review]

File: osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs:42]: If this is intended as a method for all rulesets to use, do you think we instead specify explicit osu and catch branches and leave the `default` branch as simple/lossless as possible?

Comment (bdach) [reply osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs:42]: Dunno. It may be sunk cost fallacy from all the overnight sheet running but I tend to think that this is okay. The fact that we're using legacy `.osu` format and encoding slider velocity into files in one of the most opaque ways I can imagine also contributes to this feeling - like if we were redoing this to be simpler, then we'd redefine the actual way the thing is _stored_ to not require the weird division by 100.

Can wait until further work in my opinion.

Comment (smoogipoo) [reply osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs:42]: I mean it from a different angle - this method is essentially a legacy method that pollutes an interface intended for any ruleset to use.

I can imagine any ruleset consumer would use this as a shortcut for `timingPoint.BeatLength * SliderVelocityMultiplier` without understanding the consequences of this - the precision and the weird clamping behaviour.

How do you explain the existence and purpose of this method to a ruleset consumer? I believe it should be more properly indicated.

Comment (bdach) [reply osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs:42]: I see.

For one, it has to be public API, so that's that, but what'd work for you? More xmldoc? Moving to extension method? Marking as deprecated with inline disables?

Comment (peppy) [reply osu.Game/Rulesets/Objects/Types/IHasSliderVelocity.cs]: I've moved it to an extension method. I've added more xmldoc, and I've added a `throw` on non-legacy use.

File: osu.Game/Rulesets/Objects/Legacy/LegacyRulesetExtensions.cs
Comment (smoogipoo) [inline osu.Game/Rulesets/Objects/Legacy/LegacyRulesetExtensions.cs:16]: I still don't like this being an extension method. We've gone from:

`I don't see a reason why rulesets wouldn't use this method, and it'll cause precision loss`
To:
`I don't see a reason why rulesets wouldn't use this method, but it'll crash if they do`

My suggestion was very explicit: have this return `BeatLength * SliderVelocity` for non-legacy rulesets. A second option I would be okay with is this being an explicit static class call `LegacyRulesetExtensions.Get...()`.

I don't know. I don't think I'm being as unreasonable here as I'm being brought to believe. I defer to a 3rd opinion.

Comment (bdach) [reply osu.Game/Rulesets/Objects/Legacy/LegacyRulesetExtensions.cs:16]: > I would be okay with is this being an explicit static class call `LegacyRulesetExtensions.Get...()`.

This sounds like a reasonable middle ground... @peppy?

Comment (peppy) [reply osu.Game/Rulesets/Objects/Legacy/LegacyRulesetExtensions.cs:16]: Whatever works

**Review by bdach — APPROVED** [review]

**Pull request merged**
